tests/pos/i2944.scala

```diff
@@ -0,0 +1,10 @@
+trait Map2[K] {
+  def get(k: K): K = k
+  def foo: K = {
+    this match {
+      case that: Map2[b] => that.get(3.asInstanceOf[b])
```

Comment:
This test case passes my the check with my tests 👍

Reply:
Do we have a real test for this now?